SPEC.md

```diff
@@ -490,415 +677,1104 @@ An `Array` represents an ordered list of elements that are all of the same type.
+
+<details>
+<summary>
+Example: declarations.wdl
+
+```wdl
+version 1.1
+
+workflow declarations {
+  input {
+    # these "unbound" declarations are only allowed in the input section
+    File? x  # optional - defaults to None
+    Map[String, String] m  # required
+    # this is a "bound" declaration
+    String y = "abc"  
+  }
+
+  Int i = 1 + 2  # Private declarations must be bound
+
+  output {
+    Float pi = i + .14  # output declarations must also be bound
+  }
+}
+```
+</summary>
+<p>
+Example input:
+
+```json
+{
+  "declarations.m": {"a": "b"}
+}
+```
+
+Example output:
+
+```json
+{
+  "declarations.pi": 3.14
+}
+```
+</p>
+</details>
+
+A declaration may be initialized with an [expression](#expressions), which includes the ability to refer to elements that are outputs of tasks.
+
+<details>
+<summary>
+Example: task_outputs.wdl
+
+```wdl
+version 1.1
+
+task greet {
+  input {
+    String name
+  }
+  
+  command <<<
+    printf "Hello ~{name}"
+  >>>
+
+  output {
+    String greeting = read_string(stdout())
+  }
+}
+
+task count_lines {
+  input {
+    Array[String] array
+  }
+
+  command <<<
+    wc -l ~{write_lines(array)}
+  >>>
+  
+  output {
+    Int line_count = read_int(stdout())
+  }
+}
+
+workflow task_outputs {
+  call greet as x {
+    input: name="John"
+  }
+  
+  call greet as y {
+    input: name="Sarah"
+  }
+
+  Array[String] greetings = [x.greeting, y.greeting]
+  call count_lines {
+    input: array=greetings
+  }
+
+  output {
+    Int num_greetings = count_lines.line_count
+  }
+}
+```
+</summary>
+<p>
+Example input:
+
+```json
+{}
+```
+
+Example output:
+
+```json
+{
+  "task_outputs.num_greetings": 2
+}
+```
+</p>
+</details>
+
+In this example, `greetings` is undefined until both `call greet as x` and `call greet as y` have successfully completed, at which point it is assigned the result of evaluating its expression. If either of the two tasks fail, the workflow would also fail and `greetings` would never be initialized.
+
+It must be possible to organize all of the statements within a scope into a directed acyclic graph (DAG); thus, circular references between declarations are not allowed. The following example would result in an error due to the presence of a circular reference.
+
+<details>
+<summary>
+Example: circular.wdl
+
+```wdl
+version 1.1
+
+workflow circular {
+  Int i = j + 1
+  Int j = i - 2
+}
+```
+</summary>
+<p>
+Example input:
+
+```json
+{}
+```
+
+Example output:
+
+```json
+{}
+```
+
+Test config:
+
+```json
+{
+  "fail": true
+}
+```
+</p>
+</details>
+
+### Expressions
+
+```txt
+$expression = '(' $expression ')'
```

Comment:
I dont know if removing this would be considered a breaking change, but we could always link to the grammar instead of embedding this expressions

Reply:
I'm not sure what the right thing to do is. But honestly I haven't reviewed these grammar sections closely - there are probably errors or inconsistencies with the official grammar. If we keep them in the spec we should at least switch to a standard notation like BNF.

Reply:
TBH I never really liked the pseudo-grammar blocks. I thought I had removed them at some point long ago but I guess not. Perhaps I just nicked a few of them 🤷 

I would advocate for removing them and then as @jdidion said, perhaps look at providing a BNF as an appendix or something

Reply:
Looking at some grammars being presented prompted the thought -> I think someone already **has** done a BNF of WDL, or at least a partial one. Will try to dig it up, assuming it wasn't a hallucination on my part

Reply:
Aha - top of the google search `WDL BNF` - @cjllanwarne can say more to what rough edges are here: https://github.com/broadinstitute/winstanley/blob/44896516cef2dadac35e7af0fc0d5c4dc8de2ff2/src/winstanley/wdl.bnf#L4

Reply:
Nice...I'll open an issue to add that as an appendix or otherwise clone it into openwdl